pkg/controllers/override/imagepatch.go

```diff
@@ -0,0 +1,253 @@
+		if name != "" {
+			specifiedContainers.Insert(name)
+		}
+	}
+
+	// process the init containers
+	for index, container := range podSpec.InitContainers {
+		// if specified container list is empty, the image override rule applies to all containers.
+		if len(imageOverrider.ContainerNames) == 0 || specifiedContainers.Has(container.Name) {
+			imagePath, err := generateImagePathOfPodInitContainer(gvk, index)
+			if err != nil {
+				return err
+			}
+			imageValue := patchMap[imagePath]
+			if imageValue == "" {
+				imageValue = container.Image
+			}
+
+			newImageValue, err := applyImageOverride(imageValue, imageOverrider)
+			if err != nil {
+				return fmt.Errorf("failed to apply image override: %w", err)
+			}
+			patchMap[imagePath] = newImageValue
+		}
+	}
+
+	// process the containers
+	for index, container := range podSpec.Containers {
+		// if specified container list is empty, the image override rule applies to all containers.
+		if len(imageOverrider.ContainerNames) == 0 || specifiedContainers.Has(container.Name) {
+			imagePath, err := generateImagePathOfPodContainer(gvk, index)
+			if err != nil {
+				return err
+			}
+			imageValue := patchMap[imagePath]
+			if imageValue == "" {
+				imageValue = container.Image
+			}
+
+			newImageValue, err := applyImageOverride(imageValue, imageOverrider)
+			if err != nil {
+				return fmt.Errorf("failed to apply image override: %w", err)
+			}
+			patchMap[imagePath] = newImageValue
+		}
+	}
+
+	return nil
+}
+
+// applyImageOverride applies override to oldImage to generate a new value
+func applyImageOverride(oldImage string, imageOverrider *fedcorev1a1.ImageOverrider) (string, error) {
+	// parse oldImage value into different parts for easier operation
+	newImage, err := ParseImage(oldImage)
+	if err != nil {
+		return "", fmt.Errorf("parse image failed(%s), error: %w", oldImage, err)
+	}
+
+	// apply operations in image override
+	for _, operation := range imageOverrider.Operations {
+		operator, value, component := operation.Operator, operation.Value, operation.ImageComponent
+
+		// if omitted, defaults to "replace"
+		if operator == "" {
+			operator = OperatorReplace
+		}
+
+		// if "add" or "replace", operation.Value cannot be empty
+		if (operator == OperatorReplace || operator == OperatorAdd) && value == "" {
+			return "", fmt.Errorf("the add or replace operator needs value")
+		}
+
+		switch component {
+		case Registry:
+			newImage.OperateRegistry(operator, value)
+		case Repository:
+			newImage.OperateRepository(operator, value)
+		case Tag:
+			if err := newImage.OperateTag(operator, value); err != nil {
+				return "", err
+			}
+		case Digest:
+			if err := newImage.OperateDigest(operator, value); err != nil {
+				return "", err
+			}
+		default:
+			return "", fmt.Errorf("unsupported image component(%s)", component)
+		}
+	}
+
+	return newImage.String(), nil
+}
+
+func generateImagePathOfPodContainer(gvk schema.GroupVersionKind, index int) (string, error) {
```

Comment:
Please make "containers" and "initContainers" a parameter and merge the two functions into one.

Reply:
The code has been updated.